### DISCUSSION TITLE
Oauth Data Refresh called multiple times on token expire scenario

### DISCUSSION THREAD

**Discussion by balagurusurendar** [microsoft/vscode-powerquery-sdk#288] [Q&A]
I created a custom connector for my scenario that will fetch rest api to fetch data from cloud. Connector authenticated using Oauth connection. 
for Initial token fetch it's working fine it is fetching single time. when token expires on some time, refresh token calling multiple times.
light debug & found refresh tables running in parallel threads made this scenario. 

myself refreshing 20 to 30 tables at a same time. so when token expires in between them leads to this. 

what to do for make OAuth Refresh function synchronized?

my refresh function will return the json look like below
{"access_token":"1000.17631e4193b64f6f242ee3ff9411ea0b.fd5c7ab3db39169375c6fea375f77f80","scope":"****","token_type":"Bearer","expires_in":3600}

Is anything problem with this json? or connector expecting this json needs to be in different format?

Note: Individual thread refreshing the token works & each thread gets a different token from server, further api calls also successful. my issue is that calling refresh multiple times for token leads to blocking myself from fetching token for some time.

**Comment by geologyrocks** [comment]
Could you provide examples of where you are calling your refresh token method? If you use the inbuilt authentication logic, then you shouldn't need to call refresh token at all; the framework will do it for you.

See: https://jussiroine.com/2019/02/building-a-custom-connector-for-power-bi-that-supports-oauth2-to-visualize-my-wellness-data/

**Reply by balagurusurendar** [reply]
I am not calling the refresh token. Framework calling multiple times in a short span of 10 to 20 seconds. 

I am using the in build authentication method( oauth ).

**Comment by geologyrocks** [comment]
Ah ok, that is bizarre.  Could you please share your full authentication code flow - redacted and obfuscated as needed - so we can see how it's structured please?

**Reply by balagurusurendar** [reply]
I can't share much. it's my organization code. will share oauth code alone. 
[PrivateApp.txt](https://github.com/microsoft/vscode-powerquery-sdk/files/12747408/PrivateApp.txt)


this is my fiddler output
![image](https://github.com/microsoft/vscode-powerquery-sdk/assets/23736591/b2df1017-a3f4-4ba8-a948-0a018b80425b)


above calls made in short span 1 to 2 minutes. (removed some sensitive urls & data)

my connector will give navigation of multiple tables. when we select all tables & refresh this tables at a same time will lead to this issue.
these are steps
1. select all tables 
2. load everything (in this step token call will not go as we get token in initial sign in as well)
3. after some time like 50 minutes later (note: my access token valid for 60 minutes)
4. start refresh in powerbi (usually refresh takes 10 to 20 minutes based on data)
5. while access token invalidated on 60th minutes so powerbi will do a refresh
6. basically powerbi refreshing multiple tables simultaneously so token calls made to server based on number of threads powerbi refreshing data
